cacher/cacher.go

```diff
@@ -47,50 +92,76 @@ func GetFromURLOrDefaultCache(req *http.Request, cacheDir string) (string, error
 	return result, err
 }
 
-// GetFromURLOrDefaultCacheWithCallback downloads file into cacheDir and passes the content to the onData callback
+// GetFromURLOrDefaultCacheWithCallback is a function maintained for compatibility.
+// The required httpClient is created on behalf of the caller.
+// This means some of metrics registered may be less meaningful.
+func GetFromURLOrDefaultCacheWithCallback(req *http.Request, cacheDir string, onData func([]byte) error) error {
+	httpClient := &promhttp.Client{
+		Client:     http.DefaultClient,
+		Registerer: prometheus.DefaultRegisterer,
+	}
+
+	return GetFromURLOrDefaultCacheWithCallbackAndClient(req, cacheDir, onData, httpClient)
+}
+
+// GetFromURLOrDefaultCacheWithCallbackAndClient downloads file into cacheDir and passes the content to the onData callback
 //
 // If the URL is not available, attempt to read the previous response from cache
 //
 // The onData function should process the data (e.g. parsing JSON) and return error on failure. It may be called a
 // second time to process cache if the data from remote URL cannot be processed.
 //
+// The httpClient is a wrapper for the standard net/http client which registers some basic metrics.
+// We do not register errors here because they would not be meaningful.
+//
 // The function only returns remote error if both downloading from the URL and reading from existing cache fail,
 // cache-related error is only logged, not reported.
-func GetFromURLOrDefaultCacheWithCallback(req *http.Request, cacheDir string, onData func([]byte) error) error {
+func GetFromURLOrDefaultCacheWithCallbackAndClient(req *http.Request, cacheDir string, onData func([]byte) error, httpClient *promhttp.Client) error {
 
 	clogger := logger.WithFields(logger.Fields{
 		"component": "Cacher",
 		"url":       req.URL.String(),
 	})
+	cacherClient, _ := httpClient.ForRecipient("cacher")
 
 	filename := getFileNameFromURL(req.URL.String())
 	filepath := path.Join(cacheDir, filename)
 
-	httpClient := &http.Client{}
-	resp, reqErr := httpClient.Do(req)
+	requestStartTime := time.Now()
+	resp, reqErr := cacherClient.Do(req)
+	totalRequests.WithLabelValues(req.URL.String()).Inc()
+	requestDuration := time.Since(requestStartTime)
 
 	if reqErr != nil {
-		return getCache(clogger, filepath, onData, fmt.Errorf("failed to open URL: %w", reqErr))
+		// println(req.URL.String())
```

Comment:
Commented-out code should be removed. Can the TODO on the next line be resolved?